Introduce AI persona framework

dlqqq opened the pull request:
## Description

This is a major feature PR introduces an **AI persona framework** to Jupyter AI and allows multiple AI personas to be added per chat. Personas are analogous to "bot accounts" in other chat applications in Slack or Discord. 

High-level summary of changes:

- "Jupyternaut" has been redefined as a default AI persona provided by Jupyter AI. Jupyternaut will use the same interface as every other AI persona.

- **This PR changes Jupyternaut to only reply when `@`-mentioned in the chat**. All AI personas, including Jupyternaut, will only reply when `@`-mentioned. This allows users to precisely control when & which AI personas respond. 

    - This also enables future work surrounding *multi-agent collaboration (MAC)*. AI personas are able to `@`-mention each other to dispatch tasks or use agentic tools only available through another persona. 

- **This PR allows other packages to add fully-custom AI personas.** Developers can define the name, avatar, and how new messages are handled. Custom AI personas can use any AI model or AI library of their choice, and now have *full control* over how new messages are handled.

## Demo


https://github.com/user-attachments/assets/d6363cb3-bd8e-4f56-8958-8b8b3381674d


## Technical summary

Jupyter AI v2 had a concept of "personas", but these were essentially just names & avatars that could only appear when custom model providers were in use. This caused several issues for developers:

- In v2, a custom persona only appeared in the chat after the user manually switched to a custom model provider and sent a new message.
- In v2, developers were forced to use LangChain to build a custom model provider to provide a custom persona.
- In v2, developers could not change any details about "how" their custom model provider was being called. That was managed exclusively by `DefaultChatHandler` in Jupyter AI.

**This PR completely re-defines the concept of a persona in Jupyter AI v3** to address these issues.

Personas are now each defined in a class extending `BasePersona`, an abstract base class provided by Jupyter AI. A summary of the interface is as follows:

```python
class BasePersona(ABC):
    ychat: YChat
    manager: 'PersonaManager'
    config: ConfigManager
    log: Logger
    awareness: PersonaAwareness

    @property
    @abstractmethod
    def defaults(self) -> PersonaDefaults:
        pass
    
    @abstractmethod
    async def process_message(self, message: Message) -> None
        pass
```

-  Each persona instance is scoped to a single chat under `self.ychat`.
- Each persona can define its own name under `self.defaults`. `PersonaDefaults` is a Pydantic data model that allows a persona to define its own name and avatar.
    - We want to eventually offer users persona-specific configuration to allow users to change the name & avatar of each persona at runtime through the Jupyter AI settings. That's why the persona name & avatar are defined under `self.defaults`.
- Each persona fully defines how new messages are handled in `self.process_message()`. This is just a plain async function which can do anything. You can define a provider to use any AI model or AI libraries of your choice, as long as they are installed in your Python environment.
    - Writing/streaming replies should be done through the methods available on `self.ychat`, e.g. `self.ychat.add_message()`.

To help orchestrate personas for each chat, this PR also defines a new `PersonaManager` class. A new `PersonaManager` is initialized for each new `YChat` instance automatically. This class helps initialize the set of `BasePersona` instances for each chat.

To allow other packages to install personas, persona classes are loaded from the `"jupyter_ai.personas"` **entry point group (EPG)** when the server extension starts. Any third-party package can define a persona class and provide it to this EPG to add a custom persona to Jupyter AI.


### Architecture changes

- The chat handlers in v2 no longer do anything in this branch.
- The `JupyternautPersona` implementation now fully defines how Jupyternaut handles new messages, superseding `DefaultChatHandler`.

## Related issues

- Closes #1178.

## Other details

- This PR is still a WIP.
- We may add demo implementations of custom personas to the `jupyter_ai_test` package. That way, others can switch to this branch and test it via `jlpm dev-reinstall`.

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/extension.py:
```diff
@@ -302,20 +313,23 @@ async def get_chat(self, room_id: str) -> YChat:
+            # the "if not m['raw_time']" clause is necessary because every new
+            # message triggers 2 events, one with `raw_time` set to `True` and
+            # another with `raw_time` set to `False` milliseconds later.
+            # we should explore fixing this quirk in Jupyter Chat.
```
consider adding a ref to an issue in `JupyterChat`

dlqqq replied:
Good callout. There wasn't an existing issue for this, so I opened a new one: https://github.com/jupyterlab/jupyter-chat/issues/212.

Added a reference as an inline comment too.

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/extension.py:
```diff
@@ -606,3 +614,21 @@ def _init_context_providers(self):
+            # also why do uncaught exceptions produce an empty error log in Jupyter Server?
+            self.log.exception(e)
+
+        return persona_manager
```
question: doesn't the `except` clause on line 629 effectively changes the return typing to `PersonaManager | None` ?

dlqqq replied:
Nice catch. Fixed this in the latest commit. 👍

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/base_persona.py:
```diff
@@ -0,0 +1,189 @@
+from .persona_awareness import PersonaAwareness
+
+# prevents a circular import
+# `PersonaManager` types have to be surrounded in single quotes
```
just curious: any way to fix the import structure here?

dlqqq replied:
I don't think so (at least without compromising our type safety). `PersonaManager` manages instances of `BasePersona`, and `BasePersona` needs to annotate `self.manager` as being of type `PersonaManager`.

A circular dependency is inevitable here, and from my experience, Python's module resolver doesn't handle these well. I suspect that's because it doesn't have a concept of declarations like most other languages. Anyways, by nesting the import under `if TYPE_CHECKING:`, this prevents a runtime error while staying type-safe.

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/base_persona.py:
```diff
@@ -0,0 +1,189 @@
+        Must return a PersonaDefaults object that represents the default
+        settings of this persona.
+        """
+        pass
```
how about `raise NotImplementedError` if this method must be implemented by child classes?

dlqqq replied:
Abstract classes should extend the `abc.ABC` class in Python, part of the stdlib. That automatically enforces abstract method implementation in child classes:

<img width="880" alt="Screenshot 2025-04-16 at 10 26 20 AM" src="https://github.com/user-attachments/assets/025bf238-c9e7-4869-aae8-892f2312accb" />

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/base_persona.py:
```diff
@@ -0,0 +1,189 @@
+        """
+        # support streaming
+        # handle multiple processed messages concurrently (if model service allows it)
+        pass
```
same suggestion here.

dlqqq replied:
Abstract classes should extend the `abc.ABC` class in Python, part of the stdlib. That automatically enforces abstract method implementation in child classes:

<img alt="Screenshot 2025-04-16 at 10 26 20 AM" width="880" src="https://private-user-images.githubusercontent.com/44106031/434448913-025bf238-c9e7-4869-aae8-892f2312accb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4MjQ3MzYsIm5iZiI6MTc0NDgyNDQzNiwicGF0aCI6Ii80NDEwNjAzMS80MzQ0NDg5MTMtMDI1YmYyMzgtYzllNy00ODY5LWFhZTgtODkyZjIzMTJhY2NiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE2VDE3MjcxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFlYmM2YmY3NTIxZTczOTkxMDY2ZGZlMDY5NzIyYjA4YzIwOWM1Y2EwYjU5MGE4M2M0MGJiNjljMjRkZjcyMzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UM90xHNhRvfho_hUOOcjBY8u5SmjKm_bk4gjaxrc7qk">

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/base_persona.py:
```diff
@@ -0,0 +1,189 @@
+            avatar_url=self.avatar_path,
+        )
+
+    def as_user_dict(self) -> Dict[str, Any]:
```
I know it's WIP, but consider adding docstrings to the instance methods and properties

dlqqq replied:
Done! All properties & methods on `BasePersona` have docstrings now.

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/base_persona.py:
```diff
@@ -0,0 +1,189 @@
+        user = self.as_user()
+        return asdict(user)
+
+    async def forward_reply_stream(self, reply_stream: "AsyncIterator"):
```
question: do we use typings for async methods in the code base?

It would be something like `Awaitable[None]`

dlqqq replied:
Done in latest commit! Thanks for the suggestion.

dlqqq replied:
Actually the return type is just `None`. `mypy` automatically wraps return values in `Coroutine[Any, Any, *]`. This is what happens when I annotate the return type as just `None`:

<img width="777" alt="Screenshot 2025-05-19 at 3 30 47 PM" src="https://github.com/user-attachments/assets/31cfaf98-c815-4c27-9a2e-0f75842fb5ec" />

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/jupyternaut/jupyternaut.py:
```diff
@@ -0,0 +1,56 @@
+    def build_runnable(self) -> Any:
+        # TODO: support model parameters. maybe we just add it to lm_provider_params in both 2.x and 3.x
+        llm = self.config.lm_provider(**self.config.lm_provider_params)
+        runnable = JUPYTERNAUT_PROMPT_TEMPLATE | llm | StrOutputParser()
```
sanity-check: constant `JUPYTERNAUT_PROMPT_TEMPLATE` comes before `llm` in the `runnable` definition?

If yes, consider adding an inline comment to explain why that is.

dlqqq replied:
This is [LCEL](https://python.langchain.com/docs/concepts/lcel/#composition-syntax), a custom syntax implemented by LangChain. In LangChain, many of the objects are "runnables" that can be composed together using the pipe `|` operator. This line basically ensures that the LLM gets our custom prompt template, and coerces the output to a string.

To be honest, I do not know why runnable components need to be ordered in a certain way. I just know that this is the order that is expected by LangChain, with the prompt template first & output parser last. The official docs linked earlier don't seem to elaborate on it. Let me know if you find any details on this.

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/jupyternaut/jupyternaut.py:
```diff
@@ -0,0 +1,56 @@
+            get_session_history=lambda: YChatHistory(ychat=self.ychat, k=2),
+            input_messages_key="input",
+            history_messages_key="history",
+        )
```
nit: would just do `return RunnableWithMessageHistory` to avoid redefining `runnable`

dlqqq replied:
I prefer always assigning return values to a local variable to allow us to log the return value before returning. Since `runnable` is already allocated on the stack here, this convenience is free.

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/persona_awareness.py:
```diff
@@ -0,0 +1,85 @@
+
+    def set_local_state_field(self, field: str, value: Any) -> None:
+        with self.as_custom_client():
+            return self.awareness.set_local_state_field(field, value)
```
remove `return` ?

dlqqq replied:
Fixed, thanks!

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/persona_manager.py:
```diff
@@ -0,0 +1,174 @@
+                    f"  - Unable to load AI persona from entry point `{persona_ep.name}` due to an exception printed below."
+                )
+                self.log.exception(e)
+                continue
```
swallowing all the exceptions? Perhaps that's okay, but consider adding inline comment to explain why this is better than failing outright.

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/persona_manager.py:
```diff
@@ -0,0 +1,174 @@
+    def personas(self) -> dict[str, BasePersona]:
+        return self._personas
+
+    def get_mentioned_personas(self, new_message: Message) -> list[BasePersona]:
```
nit: docstring

dlqqq replied:
Done, thanks for the catch!

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/persona_manager.py:
```diff
@@ -0,0 +1,174 @@
+        mentioned_persona_names = [persona.name for persona in mentioned_personas]
+        self.log.info(
+            f"Received new user message mentioning the following personas: {mentioned_persona_names}."
+        )
```
optional: isn't this a lot of stdout, maybe `debug` ?

dlqqq replied:
I think it's good to be verbose with our logs this early in v3 development. We can address clean logging in the future, probably in the RC phase.

dlqqq replied:
Opened a new issue for you: #1339.

JGuinegagne commented on packages/jupyter-ai/jupyter_ai/personas/persona_manager.py:
```diff
@@ -0,0 +1,174 @@
+            if mentioned_id in self.personas:
+                persona_list.append(self.personas[mentioned_id])
+
+        return persona_list
```
question: any risk of duplication here?

otherwise you can dedup by using a set to buffer the `mentioned_id` or with:
```py
return list(set(persona_list))
```

dlqqq replied:
No risk of duplication since `mentioned_ids` is itself a set, i.e. each persona ID in that set is unique. We also forbid personas having identical IDs in `_init_personas()`:

```py
if persona.id in personas:
    class_name = persona.__class__.__name__
    self.log.warning(
        f"  - WARNING: Skipping persona '{persona.name}' from '{persona.__module__}' because another persona has an identical ID '{persona.id}'. "
        + f"Personas must all have unique IDs. Please rename the persona class from '{class_name}' to something unique to dismiss this warning."
    )
    continue
```

3coins commented:
@dlqqq 
While the overall concept for personas is spot on and aligns with the growing interest in multi-agent use-case, I think the resulting UX is not optimal with the opinionated `@persona` approach. Jupyter AI consumers (users and extension authors) would require more flexibility in handing the persona calls, for example in chats with a single persona and human, use of  `@persona` is redundant (and annoying). Another example is in turn-by-turn conversations where follow up questions are a very natural part of a conversation, having users add `@persona` to each of these adds friction to the chat experience. Did we consider making the [PersonaManager](https://github.com/jupyterlab/jupyter-ai/pull/1324/files#diff-0bac281a668dde69e29dc4b222df425a6dec7c2d3eb7451876e5cc6bc5cc6df6) a configurable traitlet, so extension authors can control the behavior of [route_message](https://github.com/jupyterlab/jupyter-ai/pull/1324/files#diff-0bac281a668dde69e29dc4b222df425a6dec7c2d3eb7451876e5cc6bc5cc6df6R154) method. This opens up the possibility of having another agent(LLM or service) or a set of heuristics to route the user messages, making the message routing completely natural and flexible. 

Some other implementation specific observations:
1. It seems like the current approach is to add all registered persona as users during [init](https://github.com/jupyterlab/jupyter-ai/pull/1324/files#diff-701d4daf2a659295d297b726ecf955886dbc31e5fb78fb63cba861a857ae5be8R104), regardless of whether they participated (**@persona**) in the conversation or not. This makes it harder to figure out the actual participants of the chat (this can still be done perhaps by inspecting the messages of a chat).
2. The current Persona object has the following properties that help identify it's purpose/skills - name, description. While **name** and **description** are helpful to some extent in identifying the things that this persona can do, it doesn't provide the full list of things that this persona can do, which can be useful in cases where custom PersonaManager would use that information to auto-route the user messages. Would it be useful to look at some of the properties defined by [a2a](https://google.github.io/A2A/specification/agent-card/) to define the capabilities of the persona/agent? 
3. The **system_prompt** seems to be completely bypassed for Jupyternaut, so not sure the actual purpose of this field. It seems like a Persona can handle much of the agent creation part on it's own, so not clear how having system_prompt helps external to the Persona. Is the intended use geared towards configuring the prompt via a config object?

3coins commented on packages/jupyter-ai/jupyter_ai/personas/base_persona.py:
```diff
@@ -0,0 +1,251 @@
+    # base class methods, available to subclasses.
+    ################################################
+    @property
+    def id(self) -> str:
```
Is this safe?
https://docs.python.org/3/library/functions.html#id

dlqqq replied:
Yes, language built-ins are not reserved keywords in Python. It's safe to name attributes `id`, especially since calling it requires the syntax `persona.id` and not just `id` on its own.

3coins commented on packages/jupyter-ai/jupyter_ai/personas/base_persona.py:
```diff
@@ -0,0 +1,251 @@
+                    append=True,
+                )
+        except Exception as e:
+            self.log.error(
+                f"Persona '{self.name}' encountered an exception printed below when attempting to stream output."
+            )
+            self.log.exception(e)
```
Why 2 separate statements? Wouldn't `log.exception` do both?
```python
self.log.exception(
  f"Persona '{self.name}' encountered an exception printed below when attempting to stream output."
)
```

dlqqq replied:
You're right. I didn't know that `self.log.exception()` could get the exception details from the surrounding context. Fixed in latest commit.

3coins commented on packages/jupyter-ai/jupyter_ai/personas/persona_manager.py:
```diff
@@ -0,0 +1,174 @@
+                    f"  - Loaded AI persona class '{class_name}' from '{class_module}' using entry point '{persona_ep.name}'."
+                )
+            except Exception as e:
+                self.log.error(
+                    f"  - Unable to load AI persona from entry point `{persona_ep.name}` due to an exception printed below."
+                )
+                self.log.exception(e)
```
Why 2 separate statements? Wouldn't log.exception do both?

dlqqq replied:
Fixed too, see above.

3coins commented on packages/jupyter-ai/jupyter_ai/personas/persona_manager.py:
```diff
@@ -0,0 +1,174 @@
+                )
+                personas[persona.id] = persona
+
+        elapsed_time_ms = (time_ns() - start_time_ns) // 1000
```
Should this have 1_000_000 instead of 1000?

dlqqq replied:
YES, great catch!! Thank you.

fperez commented:
Thanks for this work, I'm excited to start testing it and exploring its use for some of my needs (research, teaching & industry). Given that this introduces a fairly major revamp of the current experience, I wonder how hard it would be to provide an easy path for users like me, who aren't 100% in-the-weeds of JLab extension development, to test it and provide feedback. This particular effort, I think, could really use some kicking-the-tires by users (like me) to provide real-world input on the UX, flow, extensibility, etc.

You @dlqqq mention above an easy way of testing it - I'd be grateful if that was possible (or if it already is, LMK perhaps with a minimal explanation for a non-jlab-dev-expert?), and would be delighted to dogfood this and help iterate.

dlqqq commented:
@JGuinegagne @3coins I've addressed all of the feedback. Thank you both for the detailed code review! I'll fix the CI errors shortly once I'm able to see the failures again.

@3coins Thank you also for the high-level comments & for putting so much thought into how this will affect the UX. Let me respond:

> I think the resulting UX is not optimal with the opinionated @persona approach. Jupyter AI consumers (users and extension authors) would require more flexibility in handing the persona calls, for example in chats with a single persona and human, use of @persona is redundant (and annoying). Another example is in turn-by-turn conversations where follow up questions are a very natural part of a conversation, having users add @persona to each of these adds friction to the chat experience.

I agree 100%. Brian & I have discussed 2 different approaches for reducing the overhead added by this while using multiple personas:

1. Implement direct messages (DMs) in Jupyter Chat to allow you to converse w/ a single AI persona directly, dropping the need for `@`-mentioning.

2. Implement threads in Jupyter Chat. We would allow users to reply directly to an AI persona response & create a new thread, just like Slack & Discord. Inside a thread, only the original AI persona will reply, so `@`-mentioning can also be dropped there.

> Did we consider making the PersonaManager a configurable traitlet

Right now, functionality is the top priority as we are trying to make sure v3.0.0 gets released in a timely fashion. Brian had suggested that we should aggressively de-prioritize anything that can be done in a future minor release, including adding configurables.

> 1. [...] This makes it harder to figure out the actual participants of the chat

I don't know of any chat application that provides this capability, so I'm not sure if this feature is of interest to JAI users. For example, in Slack, Discord, and iMessage, the list of users is just whoever's been added to that chat. Most users added to a large channel / group chat may never post a message, yet they still appear in the users panel.

> 2. [...] Might be useful to look at some of the properties defined by [a2a](https://google.github.io/A2A/specification/agent-card/) to define the capabilities of the persona/agent?

These are great ideas, but multi-agent collaboration will have to come after we implement single-agent collaboration (e.g. getting a single agent to handle queries & use tools).

> 3. The system_prompt seems to be completely bypassed for Jupyternaut, so not sure the actual purpose of this field. It seems like a Persona can handle much of the agent creation part on it's own, so not clear how having system_prompt helps external to the Persona. Is the intended use geared towards configuring the prompt via a config object?

Yes, we eventually want it to be configurable, which is why it is an evaluated property. However, the system prompt is hard-coded as we don't have a way of editing the system prompt through the `ConfigManager`.

dlqqq commented:
@fperez Hey Fernando! Thanks for checking in and expressing interest in helping us test this new feature. We also have some demos of this ready.

After this PR is merged, I'll open a new issue that includes instructions on how to add a custom persona. I'll be sure to mention you there!

srdas reviewed (approved):
Tested and extended, now it looks good.

dlqqq commented:
I've been asked to revert merging this PR to explore if we can backport this to `2.x` first.
